crates/accelerate/src/target_transpiler/mod.rs

```diff
@@ -0,0 +1,1285 @@
+    /// ``(0, 1), (1, 0), (0, 2), (2, 0), (0, 3), (2, 3), (3, 0), (3, 2)``. If
+    /// there are properties available for the instruction you can replace the
+    /// ``None`` value in the properties dictionary with an
+    /// :class:`~qiskit.transpiler.InstructionProperties` object. This pattern
+    /// is repeated for each :class:`~qiskit.circuit.Instruction` the target
+    /// supports.
+    ///
+    /// Args:
+    ///     instruction (Union[qiskit.circuit.Instruction, Type[qiskit.circuit.Instruction]]):
+    ///         The operation object to add to the map. If it's parameterized any value
+    ///         of the parameter can be set. Optionally for variable width
+    ///         instructions (such as control flow operations such as :class:`~.ForLoop` or
+    ///         :class:`~MCXGate`) you can specify the class. If the class is specified than the
+    ///         ``name`` argument must be specified. When a class is used the gate is treated as global
+    ///         and not having any properties set.
+    ///     properties (dict): A dictionary of qarg entries to an
+    ///         :class:`~qiskit.transpiler.InstructionProperties` object for that
+    ///         instruction implementation on the backend. Properties are optional
+    ///         for any instruction implementation, if there are no
+    ///         :class:`~qiskit.transpiler.InstructionProperties` available for the
+    ///         backend the value can be None. If there are no constraints on the
+    ///         instruction (as in a noiseless/ideal simulation) this can be set to
+    ///         ``{None, None}`` which will indicate it runs on all qubits (or all
+    ///         available permutations of qubits for multi-qubit gates). The first
+    ///         ``None`` indicates it applies to all qubits and the second ``None``
+    ///         indicates there are no
+    ///         :class:`~qiskit.transpiler.InstructionProperties` for the
+    ///         instruction. By default, if properties is not set it is equivalent to
+    ///         passing ``{None: None}``.
+    ///     name (str): An optional name to use for identifying the instruction. If not
+    ///         specified the :attr:`~qiskit.circuit.Instruction.name` attribute
+    ///         of ``gate`` will be used. All gates in the ``Target`` need unique
+    ///         names. Backends can differentiate between different
+    ///         parameterization of a single gate by providing a unique name for
+    ///         each (e.g. `"rx30"`, `"rx60", ``"rx90"`` similar to the example in the
+    ///         documentation for the :class:`~qiskit.transpiler.Target` class).
+    /// Raises:
+    ///     AttributeError: If gate is already in map
+    ///     TranspilerError: If an operation class is passed in for ``instruction`` and no name
+    ///         is specified or ``properties`` is set.
+    #[pyo3(signature = (instruction, name, properties=None))]
+    fn add_instruction(
```

Comment:
I'm wondering if we want to make this different somehow. I was initially confused because the signature here is different from the python space methods. and I thought at first this was being exposed to end users (which would have been a backwards incompatible change). This might just be an issue with my review since I'm trying to keep too much state in my head at once.

Reply:
It's a little different from the python one because this function is overloaded in the python side. Whereas in python providing a name could be skipped, here it won't be needed due to Python already taking care of that already. Since we spoke of just making the Target only modifiable through Python and just having its data be accessible in rust. The `add_instruction` and `update_instruction_properties` methods are only exposed to python and not rust. And there are extra checks performed in python before they're eventually sent here.